### PR TITLE
Additional peggy smart contract testing for better code coverage.

### PR DESCRIPTION
**Pull request by khdegraaf**
A bunch of additional testing of peggy smart contracts, following Elliot's test list, as well as other code paths and error conditions.

### CODE REVIEW

**Comment by ElliotFriedman** [comment]
Hi Kevin, nice work here. This will help us ensure heightened security of our contracts going forward. 

I have a pretty large outstanding PR here https://github.com/Sifchain/sifnode/pull/442, I think we should wait on merging this in until my PR is in, then push this one through. My PR has some large architectural changes, so you will no longer call the valset contract for valset functionality as the valset and oracle contract have been merged with the cosmos bridge contract.

**Comment by khdegraaf** [comment]
OK, no problem.  I'll review and tweak after you merge.

**Review by banshee — DISMISSED** [review]

File: .github/workflows/smart-contracts.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: Smart Contracts
 
 on:
   pull_request:
-    paths: ['smart-contracts/**']
+    paths: ['.github/**,smart-contracts/**']
```
Comment (utx0) [inline .github/workflows/smart-contracts.yml:5]: why the `.github/**` ?

**Review by utx0 — COMMENTED** [review]
just a minor question from me

File: .github/workflows/smart-contracts.yml (hunk 1 above)
Comment (banshee) [reply .github/workflows/smart-contracts.yml:5]: That was from me, and I added it after I made a mistake in the yml file itself:

```
commit d144d9e1501a91648ef16452e864d1f40f1a9518 (feature/code-cov-contracts)
Author: James Moore <james@sifchain.finance>
Date:   Tue Dec 29 15:27:54 2020 -0800

    Run github actions when you change the actions file themselves

diff --git a/.github/workflows/smart-contracts.yml b/.github/workflows/smart-contracts.yml
index 50506a98..df3417ca 100644
--- a/.github/workflows/smart-contracts.yml
+++ b/.github/workflows/smart-contracts.yml
@@ -2,7 +2,7 @@ name: Smart Contracts
 
 on:
   pull_request:
-    paths: ['smart-contracts/**']
+    paths: ['.github/**,smart-contracts/**']
 
 jobs:
   test:

commit 47cf308adc76183c73024140d29bc107972bf520
Author: James Moore <james@sifchain.finance>
Date:   Tue Dec 29 15:26:21 2020 -0800

    Fixed dash

diff --git a/.github/workflows/smart-contracts.yml b/.github/workflows/smart-contracts.yml
index 5d720030..50506a98 100644
--- a/.github/workflows/smart-contracts.yml
+++ b/.github/workflows/smart-contracts.yml
@@ -32,6 +32,4 @@ jobs:
         run: npm run test
 
       - name: Test Code Coverage
-      - run: npm run coverage
-
-
+        run: npm run coverage


```

Comment (khdegraaf) [reply .github/workflows/smart-contracts.yml]: In order to try to let changes in github actions to run stuff.  Backing it out for now as syntax has problems.

File: smart-contracts/test/test_end_to_end.js [hunk 2]
```diff
@@ -189,6 +197,330 @@ contract("CosmosBridge", function (accounts) {
...
+      const postRecipientBalance2 = bigInt(
+          String(await web3.eth.getBalance(userSeven))
+      );
+
+      var expectedBalance2 = bigInt(String(priorRecipientBalance)).plus(
+          String(this.amountWei)
+      );
+
+      const receivedFunds2 = expectedBalance.equals(postRecipientBalance);
+      receivedFunds.should.be.equal(true);
```
Comment (banshee) [inline smart-contracts/test/test_end_to_end.js:295]: Should this be receivedFunds2?

File: smart-contracts/test/test_end_to_end.js [hunk 3]
```diff
@@ -189,6 +197,330 @@ contract("CosmosBridge", function (accounts) {
...
+      const postRecipientBalance3 = bigInt(
+          String(await web3.eth.getBalance(userSeven))
+      );
+
+      var expectedBalance3 = bigInt(String(priorRecipientBalance)).plus(
+          String(this.amountWei)
+      );
+
+      const receivedFunds3 = expectedBalance.equals(postRecipientBalance);
+      receivedFunds.should.be.equal(true);
```
Comment (banshee) [inline smart-contracts/test/test_end_to_end.js:408]: receivedFunds3

File: smart-contracts/test/test_end_to_end.js [hunk 4]
```diff
@@ -189,6 +197,330 @@ contract("CosmosBridge", function (accounts) {
...
+      const postRecipientBalance4 = bigInt(
+          String(await web3.eth.getBalance(userSeven))
+      );
+
+      var expectedBalance4 = bigInt(String(priorRecipientBalance)).plus(
+          String(this.amountWei)
+      );
+
+      const receivedFunds4 = expectedBalance.equals(postRecipientBalance);
+      receivedFunds.should.be.equal(true);
```
Comment (banshee) [inline smart-contracts/test/test_end_to_end.js:521]: receivedFunds4

**Review by banshee — CHANGES_REQUESTED** [review]

File: smart-contracts/test/test_end_to_end.js (hunk 4 above)
Comment (khdegraaf) [reply smart-contracts/test/test_end_to_end.js]: fixed.

File: smart-contracts/test/test_end_to_end.js (hunk 3 above)
Comment (khdegraaf) [reply smart-contracts/test/test_end_to_end.js]: fixed

File: smart-contracts/test/test_end_to_end.js (hunk 2 above)
Comment (khdegraaf) [reply smart-contracts/test/test_end_to_end.js]: fixed.

**Review by banshee — APPROVED** [review]

File: .github/workflows/smart-contracts.yml (hunk 1 above)
Comment (utx0) [reply .github/workflows/smart-contracts.yml]: oh I get it. Thanks

**Pull request merged**